File: utils/dsa-cli/solutions/hill-climbing-search.js


const DEBUG = false;



function heuristic(a, b) {
    return Math.abs(a.x - b.x) + Math.abs(a.y - b.y);
}




/**
 * Reconstructs the path from the cameFrom map<Node, Node> by shifting backwards from the goal node to the start node
 * @param {Map<Node, Node>} cameFrom
 * @param {Node} start
 * @param {Node} goal
 * @returns {Node[]} path taken.
 */
function reconstructPath(cameFrom, start, goal) {
    let current = goal;
    const path = [current];
    while (current !== start) {
        current = cameFrom.get(current);
        path.unshift(current);
    }
    return path;
}


class Node {
    constructor(id, x, y) {
        this.id = id;
        this.x = x;
        this.y = y;
    }
}

class Graph {
    constructor() {
        this.edges = {};
    }

    addEdge(node, neighbor) {
        if (!this.edges[node.id]) {
            this.edges[node.id] = [];
        }
        this.edges[node.id].push(neighbor);
    }

    neighbors(node) {
        return this.edges[node.id];
    }

    cost(nodeA, nodeB) {
        return 1; // Assuming uniform cost for this example
    }

    dfs(currentNode, visitedNodes = new Set()) {
        visitedNodes.add(currentNode);

        if (!this.edges[currentNode.id]) {
            if (DEBUG) console.log("No neighbors (leaf node)");
            // console.log("Path", visitedNodes);
            printPath(visitedNodes)
            return;
        } // No neighbors (leaf node)

        this.neighbors(currentNode).forEach(neighbor => {
            if (!visitedNodes.has(neighbor)) {
                this.dfs(neighbor, new Set(visitedNodes));
            }
        });
    }
}


class HillClimbingSearch {

    /**
     * 
     * @param {Graph} graph a graph of nodes
     * @param {Node} start start node
     * @param {Node} goal end goal node
     * @returns {Dict} path, cost, count_searches, formatted_path, exploration_path
     */
    hillClimbing(graph, start, goal) {
        let current = start;
        let count_searches = 0;
        let exploration_path = [];

        const cameFrom = new Map();
        cameFrom.set(start, null);

        while (current !== goal) {
            count_searches += 1;
            exploration_path.push(current);

            const neighbors = graph.neighbors(current);
            if (neighbors?.length === 0) {
                break;
            }

            let bestNext = null;
            let bestHeuristic = Infinity;
            console.log("neighbors", neighbors)
            neighbors.forEach(next => {
                if (!cameFrom.has(next)) {
                    const nextHeuristic = heuristic(goal, next);
                    if (nextHeuristic < bestHeuristic) {
                        bestHeuristic = nextHeuristic;
                        bestNext = next;
                    }
                }
            });

            if (!bestNext || bestHeuristic >= heuristic(goal, current)) {
                throw new Error('The algorithm got stuck and could not find a solution.');
            }

            cameFrom.set(bestNext, current);
            current = bestNext;
        }

        return {
            path: reconstructPath(cameFrom, start, goal),
            cost: null,
            count_searches: count_searches,
            formatted_path: reconstructPath(cameFrom, start, goal).map(node => node.id).join(' -> '),
            exploration_path: exploration_path,
        };
    }



    solve(graph, start, goal) {
        return this.hillClimbing(graph, start, goal);
    }
}

module.exports = { Problem: HillClimbingSearch };

File: utils/data/test_code/hw5/solutions/hill-climbing-search.py
from functions import reconstructPath, getNodesNamesAsList, build_test_tree


def hill_climbing_search(tree, start, goal):
    queue = [start]
    count_searches = 0
    exploration_path = []
    queue_snapshot = []
    visited = set()
    cameFrom = {start: None}
    current = start
    did_find_goal = False
    usable_neighbors_count = 1
    

    while current != goal and usable_neighbors_count > 0:
        count_searches += 1

        if current in visited:
            continue

        visited.add(current)
        exploration_path.append(current.name)
        queue_snapshot.append(getNodesNamesAsList(queue))
        print("Current: " + str(current), "len(queue): " + str(len(queue)))

        if current.name == goal.name:
            print("Found goal!", goal.name)
            did_find_goal = True
            break
        
        bestNext = None
        bestHeuristic = 1000000
        usable_neighbors_count = 0

        for neighbor in current.neighbors:
            if neighbor not in visited:
                usable_neighbors_count += 1
                heuristic = current.neighbors[neighbor]
                if heuristic < bestHeuristic:
                    print("+ Found better neighbor: " + str(neighbor) + " with heuristic " + str(heuristic) + " from " + str(current))
                    bestHeuristic = heuristic
                    bestNext = neighbor

        if bestNext is not None:
            print("Choosing to travel to " + str(bestNext) + " with heuristic " + str(bestHeuristic) + " from " + str(current))
            queue.append(bestNext)
            cameFrom[bestNext] = current
        else:
            print("No usable neighbors @" + str(current) + ", neighbors are: " )
            print(tree)

        current = bestNext

    return {
        "count_searches": count_searches,
        "exploration_path": exploration_path,
        "queue_snapshot": queue_snapshot,
        "goal_found": did_find_goal,
    }




tree_data = build_test_tree()
print(hill_climbing_search(tree_data["tree"], tree_data["start"], tree_data["goal"]))



